File: src/models/narrative.py
"""Narrative data structures for the Rolling Context system."""

from pydantic import BaseModel, Field
from typing import Literal
from enum import IntEnum


class Phase(IntEnum):
    """Story phases controlling pacing and stakes."""
    HOOK = 1         # Turns 1-5: High survivability, establish mystery
    INVESTIGATION = 2  # Turns 6-20: Fog of War, clue gathering
    CLIMAX = 3       # Turn 20+: Forced confrontation, max danger


class TruthSeed(BaseModel):
    """The Bible - immutable core truth of this adventure.

    This is the hidden foundation that grounds all generated narrative.
    Never revealed to player, but guides every AI response.
    """
    villain: str = Field(description="The antagonist and their role")
    motive: str = Field(description="Why the villain does what they do")
    plot: str = Field(description="The villain's scheme/plan")
    twist: str = Field(description="The hidden revelation that changes everything")
    location: str = Field(description="Where the final confrontation occurs")
    weakness: str = Field(description="How the villain can be defeated")


class Turn(BaseModel):
    """A single exchange in the conversation."""
    role: Literal["user", "model"]
    text: str
    turn_number: int = 0


class NarrativeState(BaseModel):
    """Dynamic narrative memory for the Rolling Context system."""

    # The Bible - static, set at game start
    truth_seed: TruthSeed

    # Long-term memory - compressed summary, updated every 5 turns
    long_term_memory: str = Field(
        default="The adventure begins...",
        description="Compressed summary of all events before working memory"
    )

    # Working memory - raw text of last 2 turns
    short_term_buffer: list[Turn] = Field(
        default_factory=list,
        max_length=4,  # 2 exchanges = 4 turns (user + model each)
        description="Verbatim last 2 player-AI exchanges"
    )

    # Current phase
    phase: Phase = Phase.HOOK

    def add_turn(self, role: Literal["user", "model"], text: str, turn_number: int):
        """Add a turn to working memory, maintaining max 4 entries."""
        self.short_term_buffer.append(Turn(role=role, text=text, turn_number=turn_number))
        # Keep only last 4 turns (2 complete exchanges)
        if len(self.short_term_buffer) > 4:
            self.short_term_buffer = self.short_term_buffer[-4:]

    def get_working_memory_text(self) -> str:
        """Format working memory for prompt injection."""
        if not self.short_term_buffer:
            return "[No recent events]"

        lines = []
        for turn in self.short_term_buffer:
            prefix = "PLAYER" if turn.role == "user" else "NARRATOR"
            lines.append(f"{prefix}: {turn.text}")
        return "\n\n".join(lines)

File: src/models/narrative.ts
/**
 * Narrative data structures for the Rolling Context system.
 */

import { z } from 'zod';

/**
 * Story phases controlling pacing and stakes.
 */
export enum Phase {
  HOOK = 1,          // Turns 1-5: High survivability, establish mystery
  INVESTIGATION = 2, // Turns 6-20: Fog of War, clue gathering
  CLIMAX = 3,        // Turn 20+: Forced confrontation, max danger
}

/**
 * The Bible - immutable core truth of this adventure.
 * Never revealed to player, but guides every AI response.
 */
export const TruthSeedSchema = z.object({
  villain: z.string().describe('The antagonist and their role'),
  motive: z.string().describe('Why the villain does what they do'),
  plot: z.string().describe("The villain's scheme/plan"),
  twist: z.string().describe('The hidden revelation that changes everything'),
  location: z.string().describe('Where the final confrontation occurs'),
  weakness: z.string().describe('How the villain can be defeated'),
});

export type TruthSeed = z.infer<typeof TruthSeedSchema>;

/**
 * Character profile generated at game start.
 */
export const CharacterProfileSchema = z.object({
  name: z.string().describe('The character name'),
  class: z.string().describe('Character archetypal class (e.g. Occultist, Grave Robber)'),
  background: z.string().describe('Brief backstory explaining why they are here'),
  traits: z.array(z.string()).describe('3-5 key personality traits'),
  appearance: z.string().describe('Visual description'),
});

export type CharacterProfile = z.infer<typeof CharacterProfileSchema>;

/**
 * World context setting the tone.
 */
export const WorldContextSchema = z.object({
  theme: z.string().describe('Core thematic elements (e.g. Cosmic Horror)'),
  tone: z.string().describe('Atmospheric descriptors'),
  setting: z.string().describe('The immediate physical setting description'),
});

export type WorldContext = z.infer<typeof WorldContextSchema>;

/**
 * Options generated for interactive setup.
 */
export const SetupOptionsSchema = z.object({
  characters: z.array(CharacterProfileSchema).describe('3 distinct character options'),
  worlds: z.array(WorldContextSchema).describe('3 distinct world settings'),
});

export type SetupOptions = z.infer<typeof SetupOptionsSchema>;

/**
 * A single exchange in the conversation.
 */
export interface Turn {
  role: 'user' | 'model';
  text: string;
  turnNumber: number;
}

/**
 * Dynamic narrative memory for the Rolling Context system.
 */
export interface NarrativeState {
  // The Bible - static, set at game start
  truthSeed: TruthSeed;

  // Long-term memory - compressed summary, updated every 5 turns
  longTermMemory: string;

  // Working memory - raw text of last 2 turns (4 entries: 2 user + 2 model)
  shortTermBuffer: Turn[];

  // Current phase
  phase: Phase;
}

/**
 * Create initial narrative state with a truth seed.
 */
export function createNarrativeState(truthSeed: TruthSeed): NarrativeState {
  return {
    truthSeed,
    longTermMemory: 'The adventure begins...',
    shortTermBuffer: [],
    phase: Phase.HOOK,
  };
}

/**
 * Add a turn to working memory, maintaining max 4 entries.
 */
export function addTurn(
  state: NarrativeState,
  role: 'user' | 'model',
  text: string,
  turnNumber: number
): void {
  state.shortTermBuffer.push({ role, text, turnNumber });
  // Keep only last 4 turns (2 complete exchanges)
  if (state.shortTermBuffer.length > 4) {
    state.shortTermBuffer = state.shortTermBuffer.slice(-4);
  }
}

/**
 * Format working memory for prompt injection.
 */
export function getWorkingMemoryText(state: NarrativeState): string {
  if (state.shortTermBuffer.length === 0) {
    return '[No recent events]';
  }

  return state.shortTermBuffer
    .map((turn) => {
      const prefix = turn.role === 'user' ? 'PLAYER' : 'NARRATOR';
      return `${prefix}: ${turn.text}`;
    })
    .join('\n\n');
}
